pkg/agent/controller/networkpolicy/fqdn.go

```diff
@@ -0,0 +1,776 @@
+		fqdnToSelectorItem:     map[string]map[fqdnSelectorItem]struct{}{},
+		selectorItemToFQDN:     map[fqdnSelectorItem]sets.String{},
+		selectorItemToRuleIDs:  map[fqdnSelectorItem]sets.String{},
+	}
+	if controller.ofClient != nil {
+		if err := controller.ofClient.NewDNSpacketInConjunction(dnsInterceptRuleID); err != nil {
+			controller.idAllocator.release(dnsInterceptRuleID)
+			return nil, fmt.Errorf("failed to install flow for DNS response interception: %w", err)
+		}
+	}
+	if dnsServerOverride != "" {
+		controller.dnsServerAddr = dnsServerOverride
+	} else {
+		host, port := os.Getenv(kubeDNSServiceHost), os.Getenv(kubeDNSServicePort)
+		if host == "" || port == "" {
+			return nil, fmt.Errorf("failed to derive DNS server from agent config or Service environment variable")
+		}
+		controller.dnsServerAddr = host + ":" + port
+	}
+	return controller, nil
+}
+
+// fqdnToSelectorItem converts a FQDN expression to a fqdnSelectorItem.
+func fqdnToSelectorItem(fqdn string) fqdnSelectorItem {
+	if strings.Contains(fqdn, "*") {
+		return fqdnSelectorItem{
+			matchRegex: toRegex(fqdn),
+		}
+	}
+	return fqdnSelectorItem{matchName: fqdn}
+}
+
+// toRegex converts a FQDN wildcard expression to the regex pattern used to
+// match FQDNs against.
+func toRegex(pattern string) string {
+	pattern = strings.TrimSpace(pattern)
+	pattern = strings.ToLower(pattern)
+
+	// Replace "." as a regex literal, since it's recogized as a separator in FQDN.
+	pattern = strings.Replace(pattern, ".", "[.]", -1)
+	// Replace "*" with ".*".
+	pattern = strings.Replace(pattern, "*", ".*", -1)
+
+	// Anchor the regex match expression.
+	return "^" + pattern + "$"
+}
+
+// setFQDNMatchSelector records a FQDN and a selectorItem matches.
+// fqdnSelectorMutex must have been acquired by the caller.
+func (f *fqdnController) setFQDNMatchSelector(fqdn string, selectorItem fqdnSelectorItem) {
+	matchedSelectorItems, ok := f.fqdnToSelectorItem[fqdn]
+	if !ok {
+		f.fqdnToSelectorItem[fqdn] = map[fqdnSelectorItem]struct{}{
+			selectorItem: {},
+		}
+	} else {
+		matchedSelectorItems[selectorItem] = struct{}{}
+	}
+	matchedFQDNs, ok := f.selectorItemToFQDN[selectorItem]
+	if !ok {
+		f.selectorItemToFQDN[selectorItem] = sets.NewString(fqdn)
+	} else {
+		matchedFQDNs.Insert(fqdn)
+	}
+}
+
+// getIPsForFQDNSelectors retrieves the current IP addresses cached for FQDNs that
+// matches the selection criteria of a v1beta2.FQDN selector.
+func (f *fqdnController) getIPsForFQDNSelectors(fqdns []string) []net.IP {
+	f.fqdnSelectorMutex.Lock()
+	defer f.fqdnSelectorMutex.Unlock()
+	var matchedIPs []net.IP
+	for _, fqdn := range fqdns {
+		fqdnSelectorItem := fqdnToSelectorItem(fqdn)
+		fqdnsMatched, ok := f.selectorItemToFQDN[fqdnSelectorItem]
+		if !ok {
+			klog.ErrorS(nil, "FQDN selector is not known to the controller, cannot get IPs", "fqdnSelector", fqdnSelectorItem)
+			return matchedIPs
+		}
+		for fqdn := range fqdnsMatched {
+			if dnsMeta, ok := f.dnsEntryCache[fqdn]; ok {
```

Comment:
no need to change this in this PR, but I find the fact that we always use the same variable name `fqdn` to designate both a "raw" FQDN selector (e.g. "*.antrea.io") and an actual FQDN present in the cache (e.g. "foo.antrea.io") a bit confusing. @tnqn @GraysonWu am I missing something?

Reply:
I felt the same. I agree using a different name for actual FQDN will be helpful. Maybe `f.selectorItemToFQDN` should be renamed too.

Reply:
@GraysonWu do you think you could handle this in a follow-up PR? We don't want to delay this one

Reply:
Yea sure thing. I can do it.